edb/lib/ext/auth.edgeql

```diff
@@ -30,6 +30,13 @@ CREATE EXTENSION PACKAGE auth VERSION '1.0' {
         create constraint exclusive on ((.iss, .sub))
     };
 
+    create type ext::auth::PasswordCredential {
```

Comment:
Do we want to introduce an abstract parent to make it easier to get all credentials for an identity?

Reply:
Sure, good call! Maybe it also makes sense to add a computed backlink to that abstract credential type to make it even easier? Or does it make more sense to have separate backlinks to each different type of credential?

How about naming here? `LocalCredential`? `AbstractCredential`? `Credential`?

Reply:
> Maybe it also makes sense to add a computed backlink in Identity to that abstract credential type to make it even easier?

This makes me think again about the link direction stuff. What will the query look like - in what case will the user have a credential object first, and then find its identity following the current link (not the proposed backlink)?

Reply:
The user will have their email and password, and submit that via basic auth, and the server will see you're trying to make a password authenticated login, will look up the `Identity` based on the provided `email`, and then test the password hashes. So, yeah, I think they'll have the `Identity` first.

Reply:
`AbstractCredential` or `Credential` both work, go for whichever you like more.

Reply:
>  I think they'll have the Identity first.

In that case, I think a single optional link from Identity to Credential is more natural?

Or rather, with the exclusive constraint, it's actually equivalent in SQL performance, users just need to write the backlink filter?

Reply:
> Or rather, with the exclusive constraint, it's actually equivalent in SQL performance, users just need to write the backlink filter?

Hopefully there won't be any need for a "user" (e.g. application developer) to actually query the `*Credential` types: they'll just call the login endpoint with the credentials and the server will do the rest of the work. So the direction of the link here is something we mostly have to deal with.

Oh, also, the same `Identity` might have multiple local `Credential`s given that we'll likely implement forgot password, magic link, recovery code, MFA, etc. as separate `Credential`s for the same `Identity`.